packages/dev/core/src/Misc/textureTools.ts

```diff
@@ -230,6 +231,98 @@ export function FromHalfFloat(value: number): number {
+        lodPostProcess = new PostProcess("lodCube", "lodCube", ["lod", "gamma"], null, 1.0, null, Texture.NEAREST_NEAREST_MIPNEAREST, engine, false, faceDefines[face]);
+    }
+
+    if (!lodPostProcess.getEffect().isReady()) {
+        await new Promise((resolve) => {
+            lodPostProcess.getEffect().onCompileObservable.addOnce(() => {
+                resolve(0);
+            });
+        });
+    }
+
+    const rtt = new RenderTargetTexture("temp", { width: width, height: height }, scene, false);
+
+    lodPostProcess.onApply = function (effect) {
+        effect.setTexture("textureSampler", texture);
+        effect.setFloat("lod", lod);
+        effect.setBool("gamma", texture.gammaSpace);
+    };
+
+    const internalTexture = texture.getInternalTexture();
+
+    if (rtt.renderTarget && internalTexture) {
+        const samplingMode = internalTexture.samplingMode;
+        if (lod !== 0) {
+            texture.updateSamplingMode(Texture.NEAREST_NEAREST_MIPNEAREST);
+        } else {
+            texture.updateSamplingMode(Texture.NEAREST_NEAREST);
+        }
+
+        scene.postProcessManager.directRender([lodPostProcess], rtt.renderTarget, true);
+        texture.updateSamplingMode(samplingMode);
+
+        //Reading datas from WebGL
+        const bufferView = await engine.readPixels(0, 0, width, height);
+        const data = new Uint8Array(bufferView.buffer, 0, bufferView.byteLength);
+        resolve(data);
+
+        // Unbind
+        engine.unBindFramebuffer(rtt.renderTarget);
+    } else {
+        reject();
+    }
+
+    rtt.dispose();
+    lodPostProcess.dispose();
+};
+
+/**
+ * Gets the data of the specified texture by rendering it to an intermediate RGBA texture and retreiving the bytes from it.
+ * This is convienent to get 8-bit RGBA values for a texture in a GPU compressed format.
+ * @param texture the source texture
+ * @param width the width of the result, which does not have to match the source texture width
+ * @param height the height of the result, which does not have to match the source texture height
+ * @param face if the texture has multiple faces, the face index to use for the source
+ * @param lod if the texture has multiple LODs, the lod index to use for the source
+ * @returns the 8-bit texture data
+ */
+export async function GetTextureDataAsync(texture: BaseTexture, width: number, height: number, face: number = 0, lod: number = 0): Promise<Uint8Array> {
+    if (!texture.isReady() && texture._texture) {
+        await new Promise((resolve, reject) => {
+            if (texture._texture === null) {
+                reject(0);
+                return;
+            }
+            texture._texture.onLoadedObservable.addOnce(() => {
+                resolve(0);
+            });
+        });
+    }
+    return await new Promise((resolve, reject) => {
+        ProcessAsync(texture, width, height, face, lod, resolve, reject);
+    });

```

Comment:
This is still a dangling promise. I would recommend we update ProcessAsync to return the result in the promise and remove the resolve and reject callbacks.

Reply:
yeah agree, it's not fixed completely. My only concern with doing this is that previously I guess failure would be hidden and now they may crash the inspector?? So potentially a big behavior change, but I can try it out and see what happens when I throw an exception.

Reply:
If this is too much, we can fix it later.

Reply:
Ok, I refactored the code locally to just use Promise/async directly, and if I throw an exception out of this function, when I preview a texture in the inspector, the inspector just doesn't show the texture contents (it renders the texture as if it's completely transparent). No errors at that point. (which seems probably OK)

However, if I proceed to edit the texture, then a bunch of errors ensue.

So I guess if this method throws while editing the texture, it is not handled quite as gracefully.

Let me know if that's acceptable behavior; if not, I say we wrap up this PR and probably track this dangling promise as a bug?

Reply:
IMO, that's probably better behavior than if the error was ignored.